File: Go/Linked List/2807_Insert Greatest Common Divisors in Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func findGCD(a, b int) int {
	var check, limit int
	if a < b {
		limit, check = a, b
	}
	limit, check = b, a
	gcd := 0
	for i := 1; i <= limit; i++ {
		if (check%i) == 0 && (limit%i) == 0 {
			if i > gcd {
				gcd = i
			}
		}
	}
	if gcd == 0 {
		return limit
	}
	return gcd
}
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
	temp := head
	for temp.Next != nil {
		next := temp.Next
		gcd := findGCD(temp.Val, next.Val)
		newNode := &ListNode{Val: gcd, Next: next}
		temp.Next = newNode
		temp = next
		next = temp.Next
	}
	return head
}

File: C++/Linked List/2807_Insert _Greatest_Common _Divisors_in_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
    int gcd(int val1, int val2)
    {
        int gcd;
        if (val2 < val1)
        {
            int temp = val1;
            val1 = val2;
            val2 = temp;
        }
        for (int i = 1; i <= val1; i++)
            if ((val1 % i == 0) && (val2 % i == 0))
                gcd = i;
        return gcd;
    }

public:
    ListNode *insertGreatestCommonDivisors(ListNode *head)
    {
        ListNode *first = head, *next = head->next;
        while (next != NULL)
        {
            ListNode *temp = new ListNode(gcd(first->val, next->val), next);
            first->next = temp;
            first = next;
            next = next->next;
        }
        return head;
    }
};
